masu/external/notifications/aws/aws_notification_handler.py

```diff
@@ -0,0 +1,174 @@
+#
+"""AWS Notification Handler."""
+
+import json
+import logging
+import re
+
+import boto3
+
+from masu.external import AWS_SNS_HEADER_MSG_TYPE
+from masu.external.notifications.notification_interface import (NotificationInterface,
+                                                                NotificationInterfaceFilter)
+
+LOG = logging.getLogger(__name__)
+
+
+class AWSNotificationHandlerError(Exception):
+    """AWS Notification Handler error."""
+
+    pass
+
+
+# pylint: disable=too-few-public-methods
+class AWSNotificationHandler(NotificationInterface):
+    """AWS SNS notification handler."""
+
+    UUID_REGEX = '[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}'
+
+    def __init__(self, header, body, validation=True):
+        """
+        Initializer.
+
+        Args:
+            header          ([(,)])  List of header key/value pair sets
+            body            (String) String representation of the request body JSON
+            validation      (Boolean) Flag to skip message signature validation (optional)
+
+        """
+        self._headers = header
+        self._body = body
+        self._msg_type = self._get_message_type()
+        if validation:
+            self._validate_message()
+        self._confirm_subscription()
+
+    def _validate_message(self):
+        """
+        Validate incoming message.
+
+        Verify that the message payload matches the provided signature.
+
+        Args:
+            None
+
+        Returns:
+            TBD
+
+        """
+        # Validate signature of request body.  Throw exception if verification fails
+        pass
+
+    def get_billing_source(self):
+        """
+        Retrieve the name of the S3 bucket from the SNS notification.
+
+        Args:
+            None
+
+        Returns:
+            (String) Name of the billing source from the notification
+
+        Raises:
+            (NotificationInterfaceFilter): Is raised when the notification is not for
+                                           a top-level manifest file.
+
+        """
+        s3_bucket_name = None
+        if self._msg_type == 'Notification' and self._body:
+            body_dict = json.loads(self._body)
+            message_dict = json.loads(body_dict['Message'])
+
+            object_key = message_dict['Records'][0]['s3']['object']['key']
+            if object_key.endswith('Manifest.json'):
+                if re.findall(self.UUID_REGEX, object_key):
+                    msg = 'Ignoring non-toplevel manifest file: {}'.format(object_key)
+                    raise NotificationInterfaceFilter(msg)
+
+            if not object_key.endswith('Manifest.json'):
+                msg = 'Ignoring non-manifest file: {}'.format(object_key)
+                raise NotificationInterfaceFilter(msg)
+
+            # There must be a more reliable way of getting bucket name..
+            s3_bucket_name = message_dict['Records'][0]['s3']['bucket']['name']
```

Comment:
Just to respond to the comment about digging into deeply nested data...

If we need to do this often enough, it might make sense to use something like `objectpath`. See: http://objectpath.org/

Reply:
yeah we'll see.  I typically like to keep the dependencies to third party tools to a minimum.  If this proves to be a real pain point then we should look at using something like object path.